File: Problems 1-9/problem7.py
def sieve_of_eratosthenes(n = 10001):
	n = n + 1
	sieve = range(n)
	sieve[:2] = [0, 0]

	for i in xrange(2, int(n**0.5) + 1):
		if sieve[i]:
			for j in xrange(i**2, n, i):
				sieve[j] = 0

	#filter out the values that aren't set to zero
	return [p for p in sieve if p]

def find_nth_prime(n = 10001):
	upper_limit = n + 1
	primes = []

	while len(primes) < n:
		primes = sieve_of_eratosthenes(upper_limit)
		upper_limit *= 2
	
	return primes[n-1]


if __name__ == "__main__":
	print "The 10001st prime number is:", find_nth_prime()

File: Problems 1-9/problem7.c
#include <stdio.h>
#include <stdlib.h> 

unsigned int sieve_of_eratosthenes(unsigned int n);

int main()
{
	int n = 10001;

	printf("The 10001st prime number is: %u\n", sieve_of_eratosthenes(n));

	return 0;
}

unsigned int sieve_of_eratosthenes(unsigned int n)
{
	char *sieve;
	unsigned int i, upper_limit = 1000000;
	unsigned count = 0;
	
	sieve = calloc(n, sizeof *sieve); //Initialises to zero

	for (i = 2; i < upper_limit; i++)
	{
		if (!sieve[i])
		{
			unsigned int j;

			count++;

			if (count == n)
			{
				free(sieve); //we've reached our nth prime, clean up memory
				return i;
			}

			for (j = i*2; j < upper_limit; j += i)
			{
				sieve[j] = 1;
			}
		}
	}

	free(sieve);

	return 0;
}
